src/renderer/renderer.ts

```diff
@@ -52,25 +52,29 @@ declare const canvas2pdf: any;
 declare const vscode: any | null;
 declare const MINIMAP_ENABLED: boolean | undefined;
 
-type SDFGElementType = 'states' | 'nodes' | 'edges' | 'isedges';
+type SDFGElementGroup = 'states' | 'nodes' | 'edges' | 'isedges';
 // If type is explicitly set, dagre typecheck fails with integer node ids
 export type SDFGListType = any[];//{ [key: number]: DagreSDFG };
 
+export enum SDFGRendererEvent {
+    ADD_ELEMENT = 'ADD_ELEMENT',
+    QUERY_LIBNODE = 'QUERY_LIBNODE',
+}
+
 function check_valid_add_position(
-    type: string | null, foreground_elem: SDFGElement | undefined | null,
-    lib: any, _mousepos: any
+    type: SDFGElementType | null,
+    foreground_elem: SDFGElement | undefined | null, lib: any, _mousepos: any
 ): boolean {
     if (type !== null) {
         switch (type) {
-            case 'SDFGState':
+            case SDFGElementType.SDFGState:
                 return (foreground_elem instanceof NestedSDFG ||
```

Comment:
uh shouldn't this be nested SDFG?

Reply:
Why? It's checking if the element you're trying to add it to is either a nested SDFG or the top-level SDFG (so no element)